docs/design_proposals/debug-events.md

```diff
@@ -0,0 +1,187 @@
+  "result": {
+    "timestamp": "2019-10-23T18:50:38.916931Z",
+    "event": {
+      "deployEvent": {
+        "status": "In Progress"
+      }
+    },
+    "entry": "Deploy started"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:39.629826Z",
+    "event": {
+      "deployEvent": {
+        "status": "Complete"
+      }
+    },
+    "entry": "Deploy complete"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:41.004396Z",
+    "event": {
+      "debugEvent": {
+        "status": "Started",
+        "podName": "npm",
+        "containerName": "web",
+        "namespace": "default",
+        "runtime": "nodejs",
+        "configuration": "{\"name\":\"gcr.io/artifact/name/from/skaffold.yaml\",\"ports\":{\"devtools\":9229},\"runtime\":\"nodejs\",\"workingDir\":\"/app\"}"
+      }
+    },
+    "entry": "Debuggable container started pod/npm:web (default)"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:41.004551Z",
+    "event": {
+      "portEvent": {
+        "localPort": 3000,
+        "remotePort": 3000,
+        "podName": "npm",
+        "containerName": "web",
+        "namespace": "default",
+        "resourceType": "pod",
+        "resourceName": "npm"
+      }
+    },
+    "entry": "Forwarding container web to local port 3000"
+  }
+}
+{
+  "result": {
+    "timestamp": "2019-10-23T18:50:41.004792Z",
+    "event": {
+      "portEvent": {
+        "localPort": 9229,
+        "remotePort": 9229,
+        "podName": "npm",
+        "containerName": "web",
+        "namespace": "default",
+        "portName": "devtools",
+        "resourceType": "pod",
+        "resourceName": "npm"
+      }
+    },
+    "entry": "Forwarding container web to local port 9229"
+  }
+}
+```
+
+### Open Issues/Questions
+
+**\<Why is a new Debug Event necessary?\>**
+
+One alternative explored was to piggy-back the debug information into the existing port-forward events.
+This design was eliminated as:
+ 
+  - There are some language runtimes, like .NET Core, that do not use debug ports and instead
+    require a `kubectl exec` to launch a process in the remote container. 
+  - Port-forwards may be dropped and re-established
+  
+There may also be a desire for the IDEs to selectively initiate port-forwards on an as-neeeded
+basis.  This may be useful for large microservice deployments.
+
+## Implementation plan
+
+The idea here is that `skaffold debug` installs a watcher that looks and notifies of pods with
+debuggable containers, similar to how the [port-forwarding manager](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/kubernetes/portforward/pod_forwarder.go) 
+listens for pods or services.
+
+1. `skaffold debug` will set a `DebugMode` option to `true`. This value will be used in
+  `SkaffoldRunner.Dev()` to install the debug container watch manager.
+1. Add a new `DebugEvent` type to the protobuf definitions.  The status field will be one
+   of `"Started"` or `"Termninated"` (or `"Stopped"`?).
```

Comment:
I like `Terminated` here more than `Stopped`. even though `Stopped` has the opposite thing going for it, `Terminated` sounds more final.